plan/predicate_push_down.go

```diff
@@ -47,10 +49,229 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
+			isHandled[i] = true
+		case ast.EQ:
+			var (
+				col      *expression.Column
+				val      *expression.Constant
+				isColumn bool
+				bytes    []byte
+			)
+			left, ok1 := expr.Args[0].(*expression.Constant)
+			right, ok2 := expr.Args[1].(*expression.Constant)
+			if col, isColumn = expr.Args[0].(*expression.Column); ok2 && isColumn {
+				val = right
+			} else if col, isColumn = expr.Args[1].(*expression.Column); ok1 && isColumn {
+				val = left
+			} else {
+				continue
+			}
+			bytes, _ = codec.EncodeValue(bytes, types.NewStringDatum(col.FromID), types.NewIntDatum(int64(col.Position)))
+			equalities[string(bytes)] = val
+			isHandled[i] = true
+			i = -1
+		}
+	}
+	// Propagate transitive inequality predicates.
+	// e.g for conditions "a = b and c = d and a = c and g = h and b > 0 and e != 0 and g like 'abc'",
+	//     we propagate constant as the following step:
+	// 1. build multiple equality predicates(mep):
+	//    =(a, b, c, d), =(g, h).
+	// 2. extract inequality predicates between one constant and one column,
+	//    and rewrite them using the root column of a multiple equality predicate:
+	//    b > 0, e != 0, g like 'abc' ==> a > 0, g like 'abc'.
+	//    ATTENTION: here column 'e' doesn't belong to any mep, so we skip "e != 0".
+	// 3. propagate constants in these inequality predicates, and we finally get:
+	//    "a = b and c = d and a = c and e = f and g = h and e != 0 and a > 0 and b > 0 and c > 0 and d > 0 and g like 'abc' and h like 'abc' ".
+	multipleEqualities := make(map[*expression.Column]*expression.Column, 0)
+	union := func(leftExpr *expression.Column, rightExpr *expression.Column) {
+		rootOfLeftExpr, ok1 := multipleEqualities[leftExpr]
+		rootOfRightExpr, ok2 := multipleEqualities[rightExpr]
+		if !ok1 && !ok2 {
+			multipleEqualities[leftExpr] = leftExpr
+			multipleEqualities[rightExpr] = leftExpr
+		} else if ok1 && !ok2 {
+			multipleEqualities[rightExpr] = rootOfLeftExpr
+		} else if !ok1 && ok2 {
+			multipleEqualities[leftExpr] = rootOfRightExpr
+		} else if !rootOfLeftExpr.Equal(rootOfRightExpr) {
+			for k, v := range multipleEqualities {
+				if v.Equal(rootOfRightExpr) {
+					multipleEqualities[k] = v
+				}
+			}
+		}
+	}
+	for _, cond := range conditions { // build multiple equality predicates.
+		expr, ok := cond.(*expression.ScalarFunction)
+		if !ok || expr.FuncName.L != ast.EQ {
+			continue
+		}
+		left, ok1 := expr.Args[0].(*expression.Column)
+		right, ok2 := expr.Args[1].(*expression.Column)
+		if ok1 && ok2 {
+			union(left, right)
+		}
+	}
+	if len(multipleEqualities) == 0 {
+		return conditions
+	}
+	inequalityFuncs := map[string]string{
+		ast.LT:   ast.LT,
+		ast.GT:   ast.GT,
+		ast.LE:   ast.LE,
+		ast.GE:   ast.GE,
+		ast.NE:   ast.NE,
+		ast.Like: ast.Like,
+	}
+	type transitiveInEqualityPredicate map[string]map[string][]*expression.Constant // transitive inequality predicates between one column and one constant.
```

Comment:
change map[string][]*expressionConstant as a List like []*struct. It's ok to use a structure without dedup function.